File: src/linkedlist/lc_1472_design_browser_history/design_browser_history.py
class Node:
    def __init__(self, val, next=None, prev=None) -> None:
        self.val = val
        self.next = next
        self.prev = prev


class BrowserHistory:

    def __init__(self, homepage: str):
        self.head = Node(homepage)
        self.curr = self.head

    def visit(self, url: str) -> None:
        temp = self.curr
        temp2 = Node(url, None, temp)
        temp.next = temp2
        self.curr = temp2

    def back(self, steps: int) -> str:
        temp = self.curr
        while temp.prev and steps > 0:
            temp = temp.prev
            steps -= 1
        self.curr = temp
        return temp.val

    def forward(self, steps: int) -> str:
        temp = self.curr
        while temp.next and steps > 0:
            temp = temp.next
            steps -= 1
        self.curr = temp
        return temp.val

File: src/linkedlist/lc_1472_design_browser_history/DesignBrowserHistory.java
class Node {
    public String url;
    public Node prev;
    public Node next;
    Node(String url, Node prev, Node next) {
        this.url = url;
        this.prev = prev;
        this.next = next;
    }
}
class BrowserHistory {

    Node head;
    Node curr;

    public BrowserHistory(String homepage) {
        this.head = new Node(homepage, null, null);
        this.curr = head;    
    }
    
    public void visit(String url) {
        Node node = new Node(url, null, null);
        this.curr.next = node;
        node.prev = this.curr;
        this.curr = this.curr.next;
    }
    
    public String back(int steps) {
        while(this.curr.prev != null) {
            this.curr = this.curr.prev;
            steps -= 1;
            if (steps == 0)
                break;
        }
        return this.curr.url;
    }
    
    public String forward(int steps) {
        while(this.curr.next != null) {
            this.curr = this.curr.next;
            steps -= 1;
            if (steps == 0)
                break;
        }
        return this.curr.url;
    }
}